core/coreunix/add.go

```diff
@@ -201,7 +201,8 @@ func (adder *Adder) Finalize() (ipld.Node, error) {
 		return nil, err
 	}
 	var root mfs.FSNode
-	root = mr.GetDirectory()
+	rootdir := mr.GetDirectory()
+	root = rootdir
```

Comment:
You need to check that the child of the root is a directory, otherwise more potentially harder-to-find bugs could follow if you allow the root to be an `mfs.File`.

> Otherwise I have to check the assertions, which is way worse than having a variable with the right type from the beginning.

I'm failing to see the problem, you could have both, making root here explicitly as a directory and check that the child of this directory is also a directory in the posterior assignment.

Reply:
The child of the root does not need to be a directory. It's ok if it's an FSNode. The code is fine. I think this discussion is just about personal preferences.